Sketcher: Implement hints for for all drawing tools and modes (consolidates previous PRs into a single PR)

longrackslabs opened the pull request:
<!-- Summary of changes -->

This PR adds structured, context-sensitive input hints for **all Sketcher geometry creation tools** by overriding the existing `getToolHints()` method in each relevant tool handler. These hints improve guidance during tool use by showing clear instructions for each step of the sketching process.

Each tool now returns a list of `Gui::InputHint` objects describing required input (mouse/keyboard) actions, improving user experience, especially for newer users. These hints are fully integrated with the Sketcher UI and respect the tool state.

Following maintainer feedback, this version replaces earlier `updateHints()` implementations with `getToolHints()` overrides for full alignment with Sketcher tool architecture.

---

## Tools updated

This PR introduces structured hints for the following geometry tools:

- ✅ Point  
- ✅ Line  
- ✅ Polyline  
- ✅ Arc  
- ✅ Circle  
- ✅ Polygon  
- ✅ Rectangle (all construction modes)  
- ✅ Slot  
- ✅ Arc Slot  
- ✅ B-spline (all 4 construction modes)  
- ✅ Parabola  
- ✅ Arc of Ellipse  
- ✅ Arc of Hyperbola  

---

## Behavior summary

Each tool's `getToolHints()` method dynamically generates context-specific hints such as:

- `"pick first point"` (left-click)
- `"pick center point"` (left-click)
- `"right-click to finish"` (right-click)
- `"U / J to increase / decrease sides"` (keyboard)

These hints are displayed using the existing `Gui::getMainWindow()->showHints(...)` mechanism and update automatically as the tool progresses.

---

## Supersedes

This PR replaces the following individual PRs, consolidating the work into a single branch:

- #21601 — Hyperbola  
- #21600 — Ellipse  
- #21598 — Parabola  
- #21597 — B-Spline  
- #21594 — Arc Slot  
- #21593 — Slot  
- #21592 — Rectangle  
- #21589 — Circle  
- #21586 — Line  
- #21585 — Polyline (LineSet)  
- #21565 — Point  
- #21560 — Arc  

These are retained for historical discussion.

---

## Before and After

**Before:**
- No structured input hints
- Inconsistent or missing tool guidance
- Users relied on guesswork for multi-step tools

**After:**
- Dynamic hint display based on tool state
- Standardized input terminology across all tools
- Improved discoverability and usability

---

## Notes

- This work only affects input hints. Tool logic, geometry creation, and constraints are unchanged.
- No cancel hints are included, per prior maintainer guidance.
- This sets the stage for potential follow-up UX polish (e.g., mode-switch hints or constraint hinting, state machinery to handle hints, etc).

longrackslabs committed:
Implement hints for arc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

longrackslabs committed:
Change hint text to lower case per comments and style guide

longrackslabs committed:
Sketcher: Add input hint to Point tool

Adds a structured input hint ("🖱 pick point location") to the Point tool
using Gui::InputHint. Hint appears immediately upon tool activation
and uses the new `%1` icon injection pattern with MouseLeft.

longrackslabs committed:
Sketcher: Add structured input hints for Polyline (LineSet) tool

- Adds context-sensitive input hints to the Polyline (LineSet) tool using Gui::InputHint.
- Shows "pick first point" and "pick next point" with MouseLeft, and "right-click to finish" with MouseRight, updating as the tool state changes.
- Hints appear in the status area to guide users through polyline creation.

longrackslabs committed:
Cleanup 'using' and short form of syntax per PR comment

longrackslabs committed:
Cleanup 'using' and short form of syntax per PR comment

longrackslabs committed:
Add Escape Key to cancel hints

longrackslabs committed:
Add MouseRight to cancel hint

longrackslabs committed:
MouseRight and Escape Key hints for all states

longrackslabs committed:
Sketcher: Add context-sensitive input hints to Line tool

- Implements updateHints() for DrawSketchHandlerLine to show "pick first point" and "pick second point" actions with appropriate mouse icons.
- Hints update dynamically based on tool state for improved user guidance.

longrackslabs committed:
Sketcher: Add context-sensitive input hints to Circle tool

- Implements updateHints() for DrawSketchHandlerCircle to show appropriate hints for center and 3-rim construction methods.
- Hints update dynamically based on tool state to guide user actions.

longrackslabs committed:
Remove cancel hints per PR/Discord feedback discussion

longrackslabs committed:
Sketcher: Refine Arc tool input hints for construction method

- updateHints() now explicitly distinguishes between Center and ThreeRim construction methods.
- Shows accurate context-sensitive hints for both arc creation modes.
- Removes generic else for future-proofing and clarity.

longrackslabs committed:
Remove cancel hints per Discord discussion

longrackslabs committed:
Merge branch 'FreeCAD:main' into sketcher-hints-polyline

longrackslabs committed:
Refactor to use using and short form of class names per PR feedback

longrackslabs committed:
Merge branch 'FreeCAD:main' into sketcher-hints-line

longrackslabs committed:
Sketcher: Refactor Polygon tool to use new input hints mechanism

- Replaces legacy getToolHints() with updateHints() following the modern pattern by @kadet
- Ensures consistent, context-sensitive hints for Polygon tool, matching Arc, Line, and Circle tools
- Removes cancel hints per maintainer feedback

longrackslabs committed:
Sketcher: Refactor Circle tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerCircle
- Provides context-sensitive hints for both center and 3-rim construction methods
- Ensures consistency with Arc, Line, and Polygon tools

longrackslabs committed:
Sketcher: Update Rectangle tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerRectangle
- Provides context-sensitive hints for all supported rectangle construction methods
- Ensures consistency with Arc, Line, Circle, and Polygon tools

Note: All rectangle construction modes are supported and can be cycled after tool activation, even if not all are directly shown in the UI.

longrackslabs committed:
Sketcher: Refactor Slot tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerSlot
- Provides context-sensitive hints for all slot drawing steps
- Ensures consistency with other Sketcher tools

longrackslabs committed:
Sketcher: Refactor Arc Slot tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerArcSlot
- Provides context-sensitive hints for both ArcSlot and RectangleSlot construction methods
- Ensures consistency with other Sketcher tools

longrackslabs committed:
Sketcher: Refactor B-spline tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerBSpline
- Provides context-sensitive hints for all B-spline drawing steps
- Ensures consistency with other Sketcher tools

longrackslabs committed:
Sketcher: Refactor Arc of Parabola tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerArcOfParabola
- Provides context-sensitive hints for each step of arc of parabola creation
- Ensures consistency with other Sketcher tools

longrackslabs committed:
Sketcher: Refactor Arc of Ellipse tool to use new input hints mechanism

- Implements updateHints() for DrawSketchHandlerArcOfEllipse
- Provides context-sensitive hints for each step of arc of ellipse creation
- Ensures consistency with other Sketcher tools

longrackslabs committed:
Sketcher: Add input hints to Arc of Hyperbola tool

- Implements updateHints() for DrawSketchHandlerArcOfHyperbola
- Provides context-sensitive hints for each step of arc of hyperbola creation
- Ensures consistency with other Sketcher tools
- Note: This only adds hints to the existing tool code; no refactor to controller/state machine pattern

longrackslabs committed:
Sketcher: Add M key hint to polyline tool

- Adds a hint for the M key ("change mode") in the polyline (lineset) tool when drawing subsequent segments.
- Improves discoverability of cycling segment/transition modes while drawing.

longrackslabs committed:
Merge branch 'sketcher-hints-point' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-line' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-polyline' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-arc' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-circle' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-circle' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-polygon' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-rectangle' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-slot' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-ellipse' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-parabola' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-bspline' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-arc_slot' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Merge branch 'sketcher-hints-hyperbola' into sketcher-hints-all-drawing-tools

longrackslabs committed:
Restore updateHints() for Circle from sketcher-hints-circle branch

longrackslabs committed:
Refactor Sketcher tools to use getToolHints() for input hints (most tools updated, a few pending fine-tuning)

longrackslabs committed:
Sketcher: Ensure input hints update after every state change in LineSet tool

- Explicitly call updateHint() after all Mode/state transitions in DrawSketchHandlerLineSet.
- Fixes missing or stale input hints when using the polyline/lineset tool with custom state machine.
- Brings hint update behavior in line with other Sketcher drawing tools.

longrackslabs committed:
Sketcher: Refactor Slot tool to use new input hints mechanism

longrackslabs committed:
Call updateHint() after setting mode

pierreporte commented:
I haven’t followed the discussion on discord about hints. On most of the tools, you can right-click to cancel current selection, but your images don’t show that. Even the code don’t have the word “cancel”. Is this normal not to have that hint?

longrackslabs committed:
Sketcher: Add getToolHints() override to Rectangle handler for unified input hints

- Revert inccorect previous inadvertent deletions (reverted to main)
- Implements getToolHints() in DrawSketchHandlerRectangle to provide context-sensitive input hints.
- Brings Rectangle tool in line with other Sketcher handlers using the standard input hint system.
- No changes to controller or parameter adaptation logic.

longrackslabs committed:
Sketcher: Remove redundant 'click' from Point tool hint text

- Change '%1 click to place a point' to '%1 place a point'
- Mouse cursor already indicates clicking is needed
- Makes hint text more concise and focused on the action

longrackslabs committed:
- Change remove implied 'click to finish" to just 'finish'
- Mouse cursor already indicates clicking is needed
- Makes hint text more concise and focused on the action"

longrackslabs committed:
Refactor all handlers to modern compact enum pattern per PR feedback

- Use 'using enum Gui::InputHint::UserInput;' for cleaner enum usage
- Direct return statements instead of intermediate variables
- QObject::tr() for concise translation calls
- Consistent modern pattern across all handlers
- Improved readability and maintainability
- Also implmemented getToolHints() in ArcSlot.

longrackslabs committed:
Add M key hint functionality to tools with multiple construction methods:

- Circle: Center ↔ 3-point modes
- Arc: Center ↔ 3-point modes
- Rectangle: Diagonal → Center → 3-corner → Center+2-corner modes
- Line: Point+length+angle → Point+width+height → 2-point modes
- B-Spline: Control points ↔ Knots modes (+ construction-method-aware hints)

Features:
- Add 'M switch mode' hints to make mode switching discoverable
- Add doConstructionMethodChanged() template specializations to update hints
- Use generic 'switch mode' text for initial implementation (vs specific mode names)
- Maintain consistent hint patterns across all tools
- Preserve existing functionality while adding discoverability

Users can now press M to cycle through construction methods in any multi-mode
drawing tool and see the mode switching option in the hints panel.

This first implementation uses generic 'switch mode' hint to avoid the complexity of tracking mode cycles with drop downs, etc. This would be where integrating hints into the underlying state machine might help

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArc.h:
```diff
@@ -84,6 +85,45 @@ class DrawSketchHandlerArc: public DrawSketchHandlerArcBase
+            switch (state()) {
+                case SelectMode::SeekFirst:
+                    return {InputHint {QObject::tr("%1 pick arc center"), {MouseLeft}},
+                            InputHint {QObject::tr("%1 switch mode"), {KeyM}}};
```
This hint is very often repeated. I wonder if there is a way to not repeat it's definition so much? For now you can just define `const InputHint switchModeHint {QObject::tr("%1 switch mode"), {KeyM}}` on the root level of function and use it in the lists.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArcSlot.h:
```diff
@@ -93,6 +94,27 @@ class DrawSketchHandlerArcSlot: public DrawSketchHandlerArcSlotBase
+        using enum Gui::InputHint::UserInput;
+
+        switch (state()) {
+            case SelectMode::SeekFirst:
+                return {InputHint {QObject::tr("%1 pick arc slot center"), {MouseLeft}}};
+            case SelectMode::SeekSecond:
+                return {InputHint {QObject::tr("%1 pick arc slot radius"), {MouseLeft}}};
+            case SelectMode::SeekThird:
+                return {InputHint {QObject::tr("%1 pick arc slot angle"), {MouseLeft}}};
+            case SelectMode::SeekFourth:
+                return {InputHint {QObject::tr("%1 pick arc slot width"), {MouseLeft}}};
```
I don't think we need the `arc slot` here repeated, `slot` will be enough or it probably can be skipped fully.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerRectangle.h:
```diff
@@ -2075,246 +2213,6 @@ void DSHRectangleControllerBase::doEnforceControlParameters(Base::Vector2d& onSk
     }
 }
 
-template<>
-void DSHRectangleController::adaptParameters(Base::Vector2d onSketchPos)
-{
```
still shall not be removed

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerLineSet.h:
```diff
@@ -767,6 +765,23 @@ class DrawSketchHandlerLineSet: public DrawSketchHandler
+            case STATUS_SEEK_Second:
+                return {InputHint {QObject::tr("%1 pick next point"), {MouseLeft}},
+                        InputHint {QObject::tr("%1 finish"), {MouseRight}},
+                        InputHint {QObject::tr("%1 change mode"), {KeyM}}};
```
switch mode? We should be consistent. Maybe we should have some factory for common hints?

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerRectangle.h:
```diff
@@ -109,6 +109,92 @@ class DrawSketchHandlerRectangle: public DrawSketchHandlerRectangleBase
     ~DrawSketchHandlerRectangle() override = default;
 
 private:
+    // ...existing code...
+
+private:
```
:)

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerRectangle.h:
```diff
@@ -109,6 +109,92 @@ class DrawSketchHandlerRectangle: public DrawSketchHandlerRectangleBase
+                switch (state()) {
+                    case SelectMode::SeekFirst:
+                        return {InputHint {QObject::tr("%1 pick first corner"), {MouseLeft}},
+                                InputHint {QObject::tr("%1 switch mode"), {KeyM}}};
```
These repetitions should also be avoided.

kadet1090 reviewed (changes requested):
Some minor remarks, nice work!

longrackslabs committed:
Per PR review feedback, replace repeated inline 'switch mode' hint definitions with const variables:
- Define 'const InputHint switchModeHint' once per tool function
- Use variable reference in all hint lists instead of inline creation
- Reduces code duplication from 30+ repetitions to 5 clean definitions

Affected tools: Circle, Arc, Rectangle, Line, B-Spline

Improves maintainability - hint text changes only need updates in one
location per tool. Follows DRY principle and reviewer feedback.

No functional changes - same hints displayed to users.

longrackslabs committed:
Per PR feedback, remove repetitive 'arc slot' prefix from hint messages:
- 'pick arc slot center' → 'pick slot center'
- 'pick arc slot radius' → 'pick slot radius'
- 'pick arc slot angle' → 'pick slot angle'
- 'pick arc slot width' → 'pick slot width'

Context is already clear from the active tool. Follows reviewer
feedback to make hints more concise and less verbose.

Improves readability while maintaining clarity

longrackslabs committed:
per PR feedback, Change 'change mode' to 'switch mode' to match all other drawing tools.

longrackslabs committed:
remove hanging Claude

longrackslabs committed:
per PR review, restore accidentally deleted Rectangle adaptParameters method

Add back DSHRectangleController::adaptParameters template specialization
that was accidentally removed during manual file reconstruction.

This method handles parameter updates during Rectangle tool interaction,
including checkbox synchronization for M key mode switching and proper
state management when transitioning between construction methods.

Without this method, mode switching leaves the tool in an inconsistent
state with misaligned parameters and out-of-sync checkboxes.

Addresses reviewer feedback - this functionality should not be removed.

hyarion commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArcSlot.h:
```diff
@@ -93,6 +94,27 @@ class DrawSketchHandlerArcSlot: public DrawSketchHandlerArcSlotBase
     ~DrawSketchHandlerArcSlot() override = default;
 
 private:
+    // ...existing code...
```
Existing code?

longrackslabs committed:
Clean up placeholder comment per PR review feedback

longrackslabs committed:
Refactor Arc hints to declarative data-driven architecture

Inspired by early feedback regarding separation of code and data, this
refactors the Arc tool hints from functional logic to a pure declarative
table-driven system.

Key improvements:
- Pure data-driven hint lookup via static table
- Complete separation of hint data from control logic
- Elimination of nested if/switch complexity
- Establishes reusable pattern for other drawing tools
- Maintains identical functionality with a cleaner architecture

This creates a scalable template that other sketcher tools and workbenches can adopt,
demonstrating proper usage of the InputHint framework through
declarative configuration rather than imperative code."

longrackslabs committed:
Convert Point tool to declarative hint system

Apply the proven declarative pattern to Point tool:
- Replace functional getToolHints() with table lookup
- Single state tool demonstrates pattern scalability
- Maintains identical functionality with cleaner architecture

longrackslabs committed:
Convert Line tool to declarative hint system

Apply declarative pattern to Line tool with 3 construction methods:
- OnePointLengthAngle, OnePointWidthHeight, TwoPoints
- All method/state combinations properly handled
- Maintains identical functionality with cleaner architecture

kadet1090 commented:
> Are we sure the new declarative version is the preferred approach here? Personally it just seems to make the logic harder to follow, not sure I see the value in this case, but curious to see what others think.

Yes. Why does it seem harder to follow here for you? From my perspective, this is as simple as it gets given state X the hints are Y and this is the style that dominates the modern UI. It does result in some repetition (like the switch mode) but given that both hints and order of hints is important it is hard for me to see a better option.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArc.h:
```diff
@@ -423,6 +432,21 @@ class DrawSketchHandlerArc: public DrawSketchHandlerArcBase
+    using HintTable = std::vector<HintEntry>;
+
+    // Static declaration
+    static const Gui::InputHint SWITCH_MODE_HINT;
+    static const HintTable ARC_HINT_TABLE;
```
For consts we are using `camelCase` too, `UPPER_CASE` is reserved for macros.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArc.h:
```diff
@@ -887,6 +911,57 @@ void DSHArcController::addConstraints()
+}
+
+// Static member definitions
+const Gui::InputHint DrawSketchHandlerArc::SWITCH_MODE_HINT = {QObject::tr("%1 switch mode"),
+                                                               {Gui::InputHint::UserInput::KeyM}};
+
+const DrawSketchHandlerArc::HintTable DrawSketchHandlerArc::ARC_HINT_TABLE = {
+    // Center method
+    {ConstructionMethod::Center,
+     SelectMode::SeekFirst,
+     {{QObject::tr("%1 pick arc center"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {ConstructionMethod::Center,
+     SelectMode::SeekSecond,
+     {{QObject::tr("%1 pick arc start point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {ConstructionMethod::Center,
+     SelectMode::SeekThird,
+     {{QObject::tr("%1 pick arc end point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+
+    // ThreeRim method
+    {ConstructionMethod::ThreeRim,
+     SelectMode::SeekFirst,
+     {{QObject::tr("%1 pick first arc point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {ConstructionMethod::ThreeRim,
+     SelectMode::SeekSecond,
+     {{QObject::tr("%1 pick second arc point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {ConstructionMethod::ThreeRim,
+     SelectMode::SeekThird,
+     {{QObject::tr("%1 pick third arc point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}}};
```
This cannot be defined as const, `tr` is pretty cheap function and should not be precached as it will break translations if user changes language.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArc.h:
```diff
@@ -887,6 +911,57 @@ void DSHArcController::addConstraints()
+     {{QObject::tr("%1 pick third arc point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}}};
+
+std::list<Gui::InputHint> DrawSketchHandlerArc::lookupArcHints(ConstructionMethod method,
+                                                               SelectMode state)
+{
+    auto it = std::find_if(ARC_HINT_TABLE.begin(),
+                           ARC_HINT_TABLE.end(),
+                           [method, state](const HintEntry& entry) {
+                               return entry.method == method && entry.state == state;
+                           });
+
+    return (it != ARC_HINT_TABLE.end()) ? it->hints : std::list<Gui::InputHint> {};
+}
```
That code is much harder to follow - I think that it would be better to just create few methods - one per construction method and based on the construction method call appropriate one.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerLine.h:
```diff
@@ -796,6 +818,50 @@ void DSHLineController::addConstraints()
+    // OnePointLengthAngle (0)
+    {0,
+     0,
+     {{QObject::tr("%1 pick first point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {0,
+     1,
+     {{QObject::tr("%1 pick second point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+
+    // OnePointWidthHeight (1)
+    {1,
+     0,
+     {{QObject::tr("%1 pick first point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {1,
+     1,
+     {{QObject::tr("%1 pick second point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+
+    // TwoPoints (2)
+    {2,
+     0,
+     {{QObject::tr("%1 pick first point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}},
+    {2,
+     1,
+     {{QObject::tr("%1 pick second point"), {Gui::InputHint::UserInput::MouseLeft}},
+      SWITCH_MODE_HINT}}};
+
+std::list<Gui::InputHint> DrawSketchHandlerLine::lookupLineHints(int method, int state)
+{
+    auto it = std::find_if(LINE_HINT_TABLE.begin(),
+                           LINE_HINT_TABLE.end(),
+                           [method, state](const HintEntry& entry) {
+                               return entry.constructionMethod == method && entry.state == state;
+                           });
+
+    return (it != LINE_HINT_TABLE.end()) ? it->hints : std::list<Gui::InputHint> {};
+}
```
Same comments as above.

kadet1090 commented on src/Mod/Sketcher/Gui/DrawSketchHandlerArc.h:
```diff
@@ -423,6 +432,21 @@ class DrawSketchHandlerArc: public DrawSketchHandlerArcBase
+    using HintTable = std::vector<HintEntry>;
+
+    // Static declaration
+    static const Gui::InputHint SWITCH_MODE_HINT;
```
instead of such const you can for example define static function that will return proper hint.

longrackslabs committed:
Refine Arc tool declarative hints - address maintainer feedback

- Use camelCase naming convention (switchModeHint, getArcHintTable)
- Move tr() calls to runtime to fix translation caching issues
- Preserve declarative table structure with runtime generation
- Add structure comments for maintainability
- Maintains identical functionality with cleaner implementation

longrackslabs committed:
Refine Point tool declarative hints - address feedback

- Use camelCase naming (getPointHintTable)
- Move tr() calls to runtime to fix translation issues
- Preserve simple declarative structure
- Point tool has no construction methods, keeps single-state simplicity

longrackslabs committed:
Refine Line tool declarative hints - address feedback

- Use camelCase naming (switchModeHint, getLineHintTable)
- Move tr() calls to runtime to fix translation issues
- Preserve declarative table structure with 3 construction methods
- Maintains all functionality for OnePointLengthAngle, OnePointWidthHeight, TwoPoints modes

longrackslabs committed:
Convert Arc of Ellipse to declarative hint system

- Replace functional switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- Maintains 4-state workflow: center, axis, start, end points
- Consistent with refined declarative pattern used in Arc/Point/Line tools

longrackslabs committed:
Convert Circle to declarative hint system

- Replace nested if/switch statements with declarative table approach
- Use camelCase naming and runtime tr() calls
- Support both Center and ThreeRim construction methods
- Maintains all state transitions: center/rim points
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Polygon to declarative hint system

- Replace switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- Simple 2-state tool: center selection and rotation/size
- Preserves polygon-specific hints (U/J keys for side count)
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Slot to declarative hint system

- Replace switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- Simple 3-state tool: start point, end point, radius
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Arc Slot to declarative hint system

- Replace switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- 4-state tool: center, radius, angle, width
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Arc of Hyperbola to declarative hint system

- Replace switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- 4-state tool: center, axis, start point, end point
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Parabola to declarative hint system

- Replace switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- 4-state tool: focus point, axis, starting point, end point
- Consistent with refined declarative pattern

longrackslabs committed:
Convert B-spline to declarative hint system

- Replace nested if/switch statements with declarative table approach
- Use camelCase naming and runtime tr() calls
- Support both ControlPoints and Knots construction methods
- Preserves special finish functionality (MouseRight)
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Ellipse to declarative hint system

- Replace nested if/switch statements with declarative table approach
- Use camelCase naming and runtime tr() calls
- Support both Center and ThreeRim construction methods
- 3-state workflow for each method
- Consistent with refined declarative pattern

longrackslabs committed:
Convert LineSet to declarative hint system

- Replace switch statement with declarative table approach
- Use camelCase naming and runtime tr() calls
- 2-mode tool: first point, then multi-point with finish/switch
- Preserves finish and mode switching functionality
- Consistent with refined declarative pattern

longrackslabs committed:
Convert Rectangle to declarative hint system

- Replace nested switch statements with declarative table approach
- Use camelCase naming and runtime tr() calls
- Support all 4 construction methods: Diagonal, CenterAndCorner, ThreePoints, CenterAnd3Points
- Each method has 4-state workflow with switch mode functionality
- Consistent with refined declarative pattern

longrackslabs commented:
Are you sure you want the declarative route?

It seemed like based on earlier comments that declarative was the way you
wanted to go.

So for now I'll stop - just let me know.


On Fri, May 30, 2025 at 13:14 Kacper Donat ***@***.***> wrote:

> *kadet1090* left a comment (FreeCAD/FreeCAD#21632)
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2923389342>
>
> @longrackslabs <https://github.com/longrackslabs> you did a lot of work
> reworking everything into the declarative system, but I don't think that
> there were consensus here that it is the way to go - contrary comments
> suggested otherwise. First of all, the methods are already pretty much
> declarative as based on the state you calculate the hints that should be
> displayed - this is declarative code. It can contain ifs, switches etc but
> it still is pretty pure function of state into hints.
>
> The way that you reworked has limitations that original code does not
> have. If some settings are dependent on user settings (like for example -
> OVPs) it will be very hard to include into the "declarative" structure. The
> code is also much harder to follow now, ordering is not clear - in a way it
> becomes less declarative because you need to follow "find_if" function
> which is pretty imperative solution.
>
> Before you had pretty easy structure with switch statement and then based
> on the state returning hints. With some tools including modes - indeed that
> becomes a bit more complicated. That however could be easily solved by
> splitting that method into few simpler ones:
>
> std::list<InputHint> getToolHints() {
>     switch (mode()) {
>         case ThreePoints: return getThreePointsToolHints();
>         ...
>     }
> }
>
> std::list<InputHint> getThreePointsToolHints() {
>     switch (state()) {
>         case ThreePoints: return { ... };
>         ...
>     }
> }
>
> That way you have both code that is readable, easy to follow and if needed
> you can easily include conditional checks or any logic that is required.
> You can of course find a way to implement it within your solution but the
> code will get even more complicated and even now it is really hard to
> follow.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2923389342>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KSTFSOHY2KQ5KOYK333BC3YTAVCNFSM6AAAAAB6D53L2OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMRTGM4DSMZUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kadet1090 commented:
> > Are we sure the new declarative version is the preferred approach here? Personally it just seems to make the logic harder to follow, not sure I see the value in this case, but curious to see what others think.
> 
> ~Yes. Why does it seem harder to follow here for you? From my perspective, this is as simple as it gets given state X the hints are Y and this is the style that dominates the modern UI. It does result in some repetition (like the switch mode) but given that both hints and order of hints is important it is hard for me to see a better option.~
> 
> **edit:** ok. You were referring to latest changes with `find_if` etc. I agree, it is much harder to follow.

You mean that comment? I originally misunderstood @tritao, I thought that he commented on the implementation state that I reviewed. I have not noticed that you made changes while writing that comment. In the comments that I made to the code I also did say that the `find_if` solution is harder to follow, but indeed I could have been more explicit just like I was in the previous comment.

Sorry for causing confusion!

longrackslabs commented:
Just let me know what you’d like me to do at this point — I’ve tried to
take the feedback into account and happy to adjust if there’s a clear path
forward. Appreciate all the input so far.



On Fri, May 30, 2025 at 15:37 Kacper Donat ***@***.***> wrote:

> *kadet1090* left a comment (FreeCAD/FreeCAD#21632)
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2923662326>
>
> Are we sure the new declarative version is the preferred approach here?
> Personally it just seems to make the logic harder to follow, not sure I see
> the value in this case, but curious to see what others think.
>
> Yes. Why does it seem harder to follow here for you? From my perspective,
> this is as simple as it gets given state X the hints are Y and this is the
> style that dominates the modern UI. It does result in some repetition (like
> the switch mode) but given that both hints and order of hints is important
> it is hard for me to see a better option.
>
> *edit:* ok. You were referring to latest changes with find_if etc. I
> agree, it is much harder to follow.
>
> You mean that comment? I originally misunderstood @tritao
> <https://github.com/tritao>, I thought that he commented on the
> implementation state that I reviewed. I have not noticed that you made
> changes while writing that comment. In the comments that I made to the code
> I also did say that the find_if solution is harder to follow, but indeed
> I could have been more explicit just like I was in the previous comment.
>
> Sorry for causing confusion!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2923662326>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KQJE5PBJXFBDJJPOZT3BDMSBAVCNFSM6AAAAAB6D53L2OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMRTGY3DEMZSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kadet1090 commented:
@longrackslabs to not leave you with verbal feedback only I did rework Arc tool to something that I think should work well enough: https://gist.github.com/kadet1090/bc57fedb547ea6566eb37644cb52ac00

I consulted this already with @hyarion and @tritao and they seem to like it. This should be a good balance between having too much in one function and still retaining full capabilities that we have thanks to this being function.

longrackslabs commented:
Thanks for the response. I’ve implemented this latest structure locally,
and honestly — it feels like the weakest of the three approaches I’ve
tried. It’s neither readable, declarative, nor maintainable, and from my
perspective, it's harder to extend and more difficult to apply to new tools.

I’m proud of the earlier versions of the work, which — while not perfect —
felt reasonably consistent, maintainable, and aligned with the feedback I
had at the time. But I can’t in good faith commit this latest version —
it’s not something I believe in technically, and it doesn’t reflect the
kind of structure I’d want to support or maintain.

As a new contributor, I want to be candid: what started as a positive,
engaged experience gradually shifted into a loop of subjective feedback and
unclear direction. I followed the feedback as best I could, but without a
clear decision-maker or technical lead, it became increasingly difficult to
know what was actually expected — or who I was building for. The tipping
point for me was the comment about “checking with others who *liked* the
approach” — it suggests there’s a parallel, back-channel decision process
happening based on personal preference rather than any shared standard or
objective metric. That’s tough to work with when you're contributing in
good faith and out in the open.

I am a busy person as I am sure you all are, and so have limited time to
contribute to things like this, so I really value focused, actionable
feedback. When things started to drift into subjective design loops without
clear ownership, it became hard to keep justifying the effort to iterate
again.

I’ve appreciated the engagement, and I’ve learned a lot in the process —
but I need to step back from this one for now.

So I’d consider this PR a research spike. Feel free to take what’s useful
from the work. Let me know if you'd prefer I close it.

On Sat, May 31, 2025 at 6:55 AM Kacper Donat ***@***.***>
wrote:

> *kadet1090* left a comment (FreeCAD/FreeCAD#21632)
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2925217388>
>
> @longrackslabs <https://github.com/longrackslabs> to not leave you with
> verbal feedback only I did rework Arc tool to something that I think should
> work well enough:
> https://gist.github.com/kadet1090/bc57fedb547ea6566eb37644cb52ac00
>
> I consulted this already with @hyarion <https://github.com/hyarion> and
> @tritao <https://github.com/tritao> and they seem to like it. This should
> be a good balance between having too much in one function and still
> retaining full capabilities that we have thanks to this being function.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2925217388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KUZKUMOJMKPY6SSU533BGYEJAVCNFSM6AAAAAB6D53L2OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMRVGIYTOMZYHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kadet1090 commented:
> The tipping point for me was the comment about “checking with others who *liked* the approach” — it suggests there’s a parallel, back-channel decision process happening based on personal preference rather than any shared standard or objective metric. That’s tough to work with when you're contributing in good faith and out in the open.

@longrackslabs I'm really sorry, it was not meant to do anything on back-channel as way of avoid doing stuff in public. The reason why I did ask people outside of github first was because I was not sure if that implementation is ok and did not want to risk putting stuff here that could be perceived requesting yet another change from you, while there is no consensus among people who commented already here about it being ok. I do however understand your point and I apologize for that, I had good intentions but you are right that it might have not been the best choice. 

As for the current state of the PR - it does not have any blocking feedback left. Your implementation is ok and while there may be another, potentially better implementations (although opinions may vary) I don't think that it is reason to block this PR from being merged. If you are happy with current state we can merge this PR as-is.

longrackslabs commented:
Thanks — I really appreciate the thoughtful response and the apology. I
understand that this kind of design work is hard to navigate, and I respect
that you're trying to balance multiple perspectives.

I've reviewed the current state of the PR, and I think it's in a good
place. It seems to strike a reasonable balance between structure, clarity,
and maintainability — and should provide a workable pattern for extending
hinting across Sketcher and potentially into other workbenches. Of course,
there are other ways this could be done, but this feels like a solid
baseline to start from.

Thanks again to everyone who engaged along the way — I learned a lot, and I
appreciated the back-and-forth, even when it got a little tricky.

Ship it!

On Sat, May 31, 2025 at 1:46 PM Kacper Donat ***@***.***>
wrote:

> *kadet1090* left a comment (FreeCAD/FreeCAD#21632)
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2925712820>
>
> The tipping point for me was the comment about “checking with others who
> *liked* the approach” — it suggests there’s a parallel, back-channel
> decision process happening based on personal preference rather than any
> shared standard or objective metric. That’s tough to work with when you're
> contributing in good faith and out in the open.
>
> @longrackslabs <https://github.com/longrackslabs> I'm really sorry, it
> was not meant to do anything on back-channel as way of avoid doing stuff in
> public. The reason why I did ask people outside of github first was because
> I was not sure if that implementation is ok and did not want to risk
> putting stuff here that could be perceived requesting yet another change
> from you, while there is no consensus among people who commented already
> here about it being ok. I do however understand your point and I apologize
> for that, I had good intentions but you are right that it might have not
> been the best choice.
>
> As for the current state of the PR - it does not have any blocking
> feedback left. Your implementation is ok and while there may be another,
> potentially better implementations (although opinions may vary) I don't
> think that it is reason to block this PR from being merged. If you are
> happy with current state we can merge this PR as-is.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/21632#issuecomment-2925712820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI67KRK4TJCALF67Q7IVX33BIIL7AVCNFSM6AAAAAB6D53L2OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMRVG4YTEOBSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maxwxyz commented:
@longrackslabs the new polyline has no hints now, LMB, RMB and M for the next mode with preview of the mode (as in the task dialog dropdown) needs to be added.
